app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,517 @@
+    LocationSettingsRequest.Builder builder = new
+        LocationSettingsRequest.Builder();
+    builder.addLocationRequest(mLocationRequest);
+
+    task = LocationServices.getSettingsClient(this).checkLocationSettings(builder.build());
+
+    locationSettingsResponseBuilder();
+  }
+
+  private void locationSettingsResponseBuilder() {
+    task.addOnCompleteListener(task -> {
+      try {
+        LocationSettingsResponse response = task.getResult(ApiException.class);
+        // All location settings are satisfied. The client can initialize location
+        // requests here.
+
+        startService(ACTION_TURN_ON_AFTER_O);
+
+        //}
+      } catch (ApiException exception) {
+        switch (exception.getStatusCode()) {
+          case LocationSettingsStatusCodes.RESOLUTION_REQUIRED:
+            // Location settings are not satisfied. But could be fixed by showing the
+            // user a dialog.
+            try {
+              // Cast to a resolvable exception.
+              ResolvableApiException resolvable = (ResolvableApiException) exception;
+              // Show the dialog by calling startResolutionForResult(),
+              // and check the result in onActivityResult().
+              resolvable.startResolutionForResult(
+                  ZimHostActivity.this,
+                  LOCATION_SETTINGS_PERMISSION_RESULT);
+            } catch (IntentSender.SendIntentException e) {
+              // Ignore the error.
+            } catch (ClassCastException e) {
+              // Ignore, should be an impossible error.
+            }
+            break;
+          case LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE:
+            // Location settings are not satisfied. However, we have no way to fix the
+            // settings so we won't show the dialog.
+            break;
+          default:
+            break;
+        }
+      }
+    });
+  }
+
+  //Advice user to turn on hotspot manually for API<26
+  void startHotspotDialog() {
+    AlertDialog.Builder builder = new AlertDialog.Builder(this, dialogStyle());
+
+    builder.setPositiveButton(getString(R.string.go_to_settings_label), (dialog, id) -> {
+      setupWifiSettingsIntent();
+    });
+
+    builder.setNeutralButton(getString(R.string.hotspot_dialog_neutral_button), (dialog, id) -> {
+      //TO DO: START SERVER WITHIN THE SERVICE.
+      //Adding a handler because sometimes hotspot can take time to turn on.
+      //TO DO: Add a progress dialog instead of handler
+      ProgressDialog progressDialog =
+          ProgressDialog.show(this, getString(R.string.progress_dialog_starting_server), "",
+              true);
+      progressDialog.show();
+      final Handler handler = new Handler();
+      handler.postDelayed(new Runnable() {
+        @Override
+        public void run() {
+          progressDialog.dismiss();
+          startService(ACTION_START_SERVER);
+        }
+      }, 7000);
```

Comment:
Why this delay?

Reply:
In devices<Oreo, after turning on the hotspot manually, his hotspot might not be enabled in the very second and after clicking on **YES, I'VE TURNED IT ON** the ip address would be null. What I observed on my device was that after I manually click to turn on my hotspot, it takes 4 to 5 seconds on average for it to be enabled. So, assuming if a user has clicked on **YES, I'VE TURNED IT ON** right after clicking his hotspot toggle, it could take 7 seconds at max(worst case).
Since we have no callback to check if the hotspot has been turned on.

Reply:
If we have to way to know when the hotspot is turned on then the best approach is to call the method which returns the IP repeatedly after 1 second. Whenever you have the IP remove the loading progress bar. If you do not get the IP within 20 or 30 seconds dismiss the progress bar and show a dialog that the hotspot is not turned on.

@macgills I would like to have your opinion on this.

Reply:
This looks good to me!

Reply:
Use Rx for doing it.

Reply:
Rx should be good, something like
```
Flowable.timer(1, SECOND)
.timeout(7, SECOND)
.map( doWeHaveAnIpAddress)
.filter()
.subscribe(
{yay we got it},
{it took over 7 seconds so we didn't get it})
```

Reply:
Also I would like to see DialogShower used here, for progressDialog as well. You can add a method `showProgress` to the class

Reply:
Fixed. df9a28c3bb828f2e6e2e4780d37ab76f8afe6ddd